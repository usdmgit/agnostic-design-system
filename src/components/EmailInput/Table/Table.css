@import '../../styles/theme/index.css';

.table-container {
  background-color: var(--table-background-color,var(--structure-color-100));
  border-bottom: var(--table-border-bottom, 1px solid var(--structure-color-300));
  border-left: var(--table-border-left, none);
  border-radius: var(--table-border-radius, var(--round-m));
  border-right: var(--table-border-right, none);
  border-top: var(--table-border-top, 1px solid var(--structure-color-300));
  width: var(--table-width, 100%);
  color: var(--table-color, var(--structure-color-800));
  font-family: var(--table-font-family, 'Inter');
  border-collapse: var(--table-container-border-colapse, separate);
  border-spacing: var(--table-container-border-spacing, 0);
  box-sizing: var(--table-container-box-sizing, border-box);
  font-size: var(--table-container-font-size, 14px);
  overflow: var(--table-container-overflow, hidden);
}

.table-container th {
  background-color: var(--table-header-background-color, var(--structure-color-200));
  border-bottom: var(--table-header-border-bottom, var(--table-border-bottom, 1px solid var(--structure-color-300)));
  border-bottom-style: var(--table-border-bottom-style, none);
  border-left: var(--table-header-border-left, var(--table-border-left, none));
  border-right: var(--table-header-border-right, var(--table-border-right, none));
  border-top: var(--table-header-border-top, var(--table-border-top, 1px solid var(--structure-color-300)));
  color: var(--table-header-color, var(--structure-color-900));
  font-weight: var(--table-header-font-weight, 300);
  padding: var(--table-padding, 10px);
}

.table-container td {
  background-color: var(--table-background-color, var(--structure-color-200));
  border-bottom: var(--table-cell-border-bottom, var(--table-border-bottom, 1px solid var(--structure-color-300)));
  border-left: var(--table-cell-border-left, var(--table-border-left, none));
  border-right: var(--table-cell-border-right, var(--table-border-right, none));
  border-top: var(--table-cell-border-top, var(--table-border-top, 1px solid var(--structure-color-300)));
  font-weight: var(--table-font-weight, 300);
  padding: var(--table-cell-padding, var(--table-padding, 10px));
  vertical-align: var(--table-container-cell-vertical-align, middle);
}

.table-container tfoot td {
  background-color: var(--table-footer-background-color, var(--structure-color-200));
  padding: var(--table-footer-padding, var(--table-padding, 10px));
}

.table-container tr:focus td {
  background-color: var(--table-row-background-color-focus, var(--structure-color-400));
}

.clickable-row {
  cursor: var(--table-clickable-row-cursor, pointer);
}
