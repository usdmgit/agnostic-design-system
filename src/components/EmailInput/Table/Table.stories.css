.custom-table {
  --table-header-border-top: 4px solid var(--error-color);
  --table-header-border-right: 4px solid var(--error-color);
  --table-header-border-left: 4px solid var(--error-color);
  --table-cell-border-left: 4px solid var(--information-color);
  --table-cell-border-right: 4px solid var(--information-color);
  --table-cell-border-bottom: 4px solid var(--warning-color);
}
