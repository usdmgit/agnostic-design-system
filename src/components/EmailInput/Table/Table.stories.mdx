# Table Component

## Type: Simple

## Table Props

|        Prop        |                   Type                   |                                                Description                                                      | Default | Mandatory |
| :----------------: | :--------------------------------------: | :-------------------------------------------------------------------------------------------------------------: | :-----: | :-------: |
|     headerList     |     array of strings or react nodes      |                     Specifies the list to be used to populate the table header                                  |    -    |   true    |
|     footerList     |     array of strings or react nodes      |                     Specifies the list to be used to populate the table footer                                  |    -    |   false   |
|     onRowClick     |                 function                 |    Specifies what happens when a user clicks in a specific row. Receives the row index and the row data.        |    -    |   false   |
|   onRowMouseEnter  |                 function                 |    Specifies what happens when a user hovers in a specific row. Receives the row index and the row data.        |    -    |   false   |
|   onRowMouseLeave  |                 function                 |  Specifies what happens when a user ceases hovering in a specific row. Receives the row index and the row data. |    -    |   false   |
|    renderFooter    |                 function                 |                                       Returns the full footer HTML structure                                    |    -    |   false   |
|    tableContent    | array of array of strings or react nodes |                  Each array inside the matrix represents a row in the table with all its content.               |    -    |   true    |
| variablesClassName |                  string                  |                              Specifies the custom class to apply styles to the table                            |    -    |   false   |

## Getting Started

```javascript
import { Table } from '@codelittinc/agnostic-design-system';

const props = {
  headerList: ['First Name', 'Last Name', 'Email'],
  tableContent: [
    ['Ana', 'Robin', 'ana@mail.com'],
    ['Beth', 'Julius', 'beth@mail.com'],
    ['Ruth', 'Katherine', 'ruth@mail.com'],
    ['Julia', 'Mark', 'julia@mail.com'],
    ['Nath', 'Sloan', 'nath@mail.com']
  ]
};

const Table = props => {
  return (
    <Table
      headerList={props.headerList}
      tableContent={props.tableContent}
      variablesClassName='my-custom-table'
    />
  );
};

export default Table;
```

## CSS Variables Available

|   Property    |      Attribute      | State |
| :-----------: | :-----------------: | :---: |
|     table     |    border-bottom    |       |
|     table     | border-bottom-style |       |
|     table     |     border-left     |       |
|     table     |    border-right     |       |
|     table     |     border-top      |       |
|     table     |        width        |       |
|     table     |        color        |       |
|     table     |     font-family     |       |
|     table     |   border-collapse   |       |
|     table     |   border-spacing    |       |
|     table     |     box-sizing      |       |
|     table     |      font-size      |       |
|     table     |     font-weight     |       |
|     table     |      overflow       |       |
|     table     |       padding       |       |
|     header    |  background-color   |       |
|     header    |    border-bottom    |       |
|     header    |     border-left     |       |
|     header    |    border-right     |       |
|     header    |     border-top      |       |
|     header    |        color        |       |
|     header    |     font-weight     |       |
|     footer    |       padding       |       |
|     cell      |    border-bottom    |       |
|     cell      |     border-left     |       |
|     cell      |    border-right     |       |
|     cell      |     border-top      |       |
|     cell      |       padding       |       |
| clickable-row |       cursor        |       |
|container-cell |   vertical-align    |       |
|    footer     |  background-color   |       |
|      row      |  background-color   | focus |
