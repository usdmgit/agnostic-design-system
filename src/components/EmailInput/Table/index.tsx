import React from 'react';
import classNames from 'classnames';

import styles from '@/components/Table/Table.css';
export interface Props {
  headerList: React.ReactNode[];
  tableContent: React.ReactNode[][];
  footerList?: React.ReactNode[];
  onRowClick?: (row: React.ReactNode, rowIndex: number) => void;
  onRowMouseEnter?: (row: React.ReactNode, rowIndex: number) => void;
  onRowMouseLeave?: (row?: React.ReactNode, rowIndex?: number) => void;
  renderFooter?: () => React.ReactNode;
  variablesClassName?: string;
}

const Table = (props: Props) => {
  const {
    headerList,
    tableContent,
    footerList,
    onRowClick,
    onRowMouseEnter,
    onRowMouseLeave,
    renderFooter,
    variablesClassName
  } = props;

  const handleRowClick = (row, rowIndex) => {
    if (onRowClick) {
      onRowClick(row, rowIndex);
    }
  };

  const handleRowMouseEnter = (row, rowIndex) => {
    if (onRowMouseEnter) {
      onRowMouseEnter(row, rowIndex);
    }
  };

  const handleRowMouseLeave = (row, rowIndex) => {
    if (onRowMouseLeave) {
      onRowMouseLeave(row, rowIndex);
    }
  };

  return (
    <table
      className={classNames(variablesClassName, styles['table-container'])}
      data-testid='table-element'
    >
      <thead data-testid='table-header'>
        <tr>
          {headerList.map((header, headerIndex) => (
            <th key={headerIndex}>{header}</th>
          ))}
        </tr>
      </thead>
      <tbody data-testid='table-body'>
        {tableContent.map((row, rowIndex) => (
          <tr
            className={onRowClick ? styles['clickable-row'] : styles['non-clickable-row']}
            tabIndex={0}
            key={rowIndex}
            onClick={() => handleRowClick(row, rowIndex)}
            onMouseEnter={() => handleRowMouseEnter(row, rowIndex)}
            onMouseLeave={() => handleRowMouseLeave(row, rowIndex)}
            onKeyPress={event => {
              if (event.key === 'Enter') {
                handleRowClick(row, rowIndex);
              }
            }}
          >
            {row.map((content, contentIndex) => (
              <td key={`${rowIndex} - ${contentIndex}`}> {content}</td>
            ))}
          </tr>
        ))}
      </tbody>
      {renderFooter
        ? renderFooter()
        : footerList && (
            <tfoot>
              <tr>
                {footerList.map((foot, footIndex) => (
                  <th key={footIndex}>{foot}</th>
                ))}
              </tr>
            </tfoot>
          )}
    </table>
  );
};

export default Table;

Table.defaultProps = {
  headerList: [],
  tableContent: [[]]
};
