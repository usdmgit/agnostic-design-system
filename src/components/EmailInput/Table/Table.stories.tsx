import React from 'react';
import Table from '@/components/Table';
import { Meta } from '@storybook/react';
import mdx from './Table.stories.mdx';
import classNames from 'classnames';
import styles from './Table.stories.css';

export default {
  title: 'Components/Table',
  component: Table,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => {
  return <Table {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  headerList: ['First Name', 'Last Name', 'Handle', 'User Type', 'Status'],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ]
};

const firstName = <p>First Name</p>;
const lastName = <p>Last Name</p>;
const type = <p>Type</p>;
const handle = <p>Type</p>;
const status = <p>Status</p>;

export const PassingElementsAsHeader = Template.bind({});
PassingElementsAsHeader.args = {
  headerList: [firstName, lastName, handle, type, status],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ]
};

export const CustomStyledTable = Template.bind({});
CustomStyledTable.args = {
  headerList: ['First Name', 'Last Name', 'Handle', 'User Type', 'Status'],
  tableContent: [
    ['John', 'Oto', '@timo', 'type1', 'Active'],
    ['Mark', 'Howard', '@mdo', 'type2', 'Inactive'],
    ['Jacob', 'Martin', '@mdo', 'type1', 'Active'],
    ['Lary', 'King', '@timo', 'type1', 'Inactive'],
    ['Helen', 'Bonham', '@timo', 'type2', 'Active'],
    ['Felix', 'Dougherty', '@timo', 'type1', 'Active'],
    ['Joseph', 'Thomas', '@timo', 'type1', 'Inactive'],
    ['Richard', 'Cardenas', '@timo', 'type1', 'Active'],
    ['Karl', 'Roy', '@twitter', 'type2', 'Active'],
    ['Sean', 'Steppe', '@timo', 'type1', 'Inactive'],
    ['Samuel', 'Marriott', '@twitter', 'type1', 'Inactive']
  ],
  variablesClassName: classNames(styles['custom-table'])
};
