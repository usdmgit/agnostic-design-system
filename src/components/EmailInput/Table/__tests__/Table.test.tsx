import React from 'react';
import Table from '@/components/Table';
import { mount } from 'enzyme';
import { render, screen } from '@testing-library/react';

const headerList = ['Name', 'Month', 'Gift'];
const body = [
  ['Lisa', 'December', 'Car'],
  ['Tina', 'June', 'Dog'],
  ['Callie', 'August', 'Flowers'],
  ['Ruth', 'September', 'Diamonds']
];

describe('Table', () => {
  it('renders correctly', () => {
    const wrapper = mount(<Table headerList={headerList} tableContent={body} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Table output', () => {
    it('Contains the correct number of header tags', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const header = screen.getByTestId('table-header').querySelectorAll('th');
      expect(header.length).toBe(headerList.length);
    });

    it('Displays the correct header content', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const header = screen.getByTestId('table-header');
      expect(header).toHaveTextContent('Name');
      expect(header).toHaveTextContent('Month');
      expect(header).toHaveTextContent('Gift');
    });

    it('Contains the correct number of rows', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const rows = screen.getByTestId('table-body').querySelectorAll('tr');
      expect(rows.length).toBe(body.length);
    });

    it('Displays the correct body content', () => {
      render(<Table headerList={headerList} tableContent={body} />);
      const header = screen.getByTestId('table-body');
      expect(header).toHaveTextContent('Lisa');
      expect(header).toHaveTextContent('December');
      expect(header).toHaveTextContent('Car');
    });
  });
});
