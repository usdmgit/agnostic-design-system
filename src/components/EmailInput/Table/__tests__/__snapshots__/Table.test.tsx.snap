// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders correctly 1`] = `
<Table
  headerList={
    Array [
      "Name",
      "Month",
      "Gift",
    ]
  }
  tableContent={
    Array [
      Array [
        "Lisa",
        "December",
        "Car",
      ],
      Array [
        "Tina",
        "June",
        "Dog",
      ],
      Array [
        "Callie",
        "August",
        "Flowers",
      ],
      Array [
        "Ruth",
        "September",
        "Diamonds",
      ],
    ]
  }
>
  <table
    className="table-container"
    data-testid="table-element"
  >
    <thead
      data-testid="table-header"
    >
      <tr>
        <th
          key="0"
        >
          Name
        </th>
        <th
          key="1"
        >
          Month
        </th>
        <th
          key="2"
        >
          Gift
        </th>
      </tr>
    </thead>
    <tbody
      data-testid="table-body"
    >
      <tr
        className="non-clickable-row"
        key="0"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        tabIndex={0}
      >
        <td
          key="0 - 0"
        >
           
          Lisa
        </td>
        <td
          key="0 - 1"
        >
           
          December
        </td>
        <td
          key="0 - 2"
        >
           
          Car
        </td>
      </tr>
      <tr
        className="non-clickable-row"
        key="1"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        tabIndex={0}
      >
        <td
          key="1 - 0"
        >
           
          Tina
        </td>
        <td
          key="1 - 1"
        >
           
          June
        </td>
        <td
          key="1 - 2"
        >
           
          Dog
        </td>
      </tr>
      <tr
        className="non-clickable-row"
        key="2"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        tabIndex={0}
      >
        <td
          key="2 - 0"
        >
           
          Callie
        </td>
        <td
          key="2 - 1"
        >
           
          August
        </td>
        <td
          key="2 - 2"
        >
           
          Flowers
        </td>
      </tr>
      <tr
        className="non-clickable-row"
        key="3"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        tabIndex={0}
      >
        <td
          key="3 - 0"
        >
           
          Ruth
        </td>
        <td
          key="3 - 1"
        >
           
          September
        </td>
        <td
          key="3 - 2"
        >
           
          Diamonds
        </td>
      </tr>
    </tbody>
  </table>
</Table>
`;
